# Cleaning Data

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

## Looking at Data

In [2]:
filepath = '../Data/world-happiness-report-2021.csv'
df21 = pd.read_csv(filepath)
df21.head()

Country name Regional indicator  Ladder score  \
0      Finland     Western Europe         7.842   
1      Denmark     Western Europe         7.620   
2  Switzerland     Western Europe         7.571   
3      Iceland     Western Europe         7.554   
4  Netherlands     Western Europe         7.464   

   Standard error of ladder score  upperwhisker  lowerwhisker  \
0                           0.032         7.904         7.780   
1                           0.035         7.687         7.552   
2                           0.036         7.643         7.500   
3                           0.059         7.670         7.438   
4                           0.027         7.518         7.410   

   Logged GDP per capita  Social support  Healthy life expectancy  \
0                 10.775           0.954                     72.0   
1                 10.933           0.954                     72.7   
2                 11.117           0.942                     74.4   
3                 10.878           0.983                     73.0   
4                 10.932           0.942                     72.4   

   Freedom to make life choices  Generosity  Perceptions of corruption  \
0                         0.949      -0.098                      0.186   
1                         0.946       0.030                      0.179   
2                         0.919       0.025                      0.292   
3                         0.955       0.160                      0.673   
4                         0.913       0.175                      0.338   

   Ladder score in Dystopia  Explained by: Log GDP per capita  \
0                      2.43                             1.446   
1                      2.43                             1.502   
2                      2.43                             1.566   
3                      2.43                             1.482   
4                      2.43                             1.501   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.106                                  0.741   
1                         1.108                                  0.763   
2                         1.079                                  0.816   
3                         1.172                                  0.772   
4                         1.079                                  0.753   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.691                     0.124   
1                                       0.686                     0.208   
2                                       0.653                     0.204   
3                                       0.698                     0.293   
4                                       0.647                     0.302   

   Explained by: Perceptions of corruption  Dystopia + residual  
0                                    0.481                3.253  
1                                    0.485                2.868  
2                                    0.413                2.839  
3                                    0.170                2.967  
4                                    0.384                2.798

In [3]:
df21.count()

Country name                                  149
Regional indicator                            149
Ladder score                                  149
Standard error of ladder score                149
upperwhisker                                  149
lowerwhisker                                  149
Logged GDP per capita                         149
Social support                                149
Healthy life expectancy                       149
Freedom to make life choices                  149
Generosity                                    149
Perceptions of corruption                     149
Ladder score in Dystopia                      149
Explained by: Log GDP per capita              149
Explained by: Social support                  149
Explained by: Healthy life expectancy         149
Explained by: Freedom to make life choices    149
Explained by: Generosity                      149
Explained by: Perceptions of corruption       149
Dystopia + residual                           149


In [4]:
filepath2 = '../Data/world-happiness-report.csv'
dfH = pd.read_csv(filepath2)
dfH.head()

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0  Afghanistan  2008        3.724               7.370           0.451   
1  Afghanistan  2009        4.402               7.540           0.552   
2  Afghanistan  2010        4.758               7.647           0.539   
3  Afghanistan  2011        3.832               7.620           0.521   
4  Afghanistan  2012        3.783               7.705           0.521   

   Healthy life expectancy at birth  Freedom to make life choices  Generosity  \
0                             50.80                         0.718       0.168   
1                             51.20                         0.679       0.190   
2                             51.60                         0.600       0.121   
3                             51.92                         0.496       0.162   
4                             52.24                         0.531       0.236   

   Perceptions of corruption  Positive affect  Negative affect  
0                      0.882            0.518            0.258  
1                      0.850            0.584            0.237  
2                      0.707            0.618            0.275  
3                      0.731            0.611            0.267  
4                      0.776            0.710            0.268

In [5]:
dfH.count()

Country name                        1949
year                                1949
Life Ladder                         1949
Log GDP per capita                  1913
Social support                      1936
Healthy life expectancy at birth    1894
Freedom to make life choices        1917
Generosity                          1860
Perceptions of corruption           1839
Positive affect                     1927
Negative affect                     1933
dtype: int64

In [6]:
filepath3 = '../Data/worldometer_coronavirus_summary_data.csv'
dfCS = pd.read_csv(filepath3)
dfCS.head()

country continent  total_confirmed  total_deaths  total_recovered  \
0  Afghanistan      Asia           120216        4962.0          71012.0   
1      Albania    Europe           132521        2456.0         130009.0   
2      Algeria    Africa           139626        3716.0          97089.0   
3      Andorra    Europe            13911         127.0          13720.0   
4       Angola    Africa            38849         900.0          33242.0   

   active_cases  serious_or_critical  total_cases_per_1m_population  \
0       44242.0               1124.0                           3021   
1          56.0                  3.0                          46100   
2       38821.0                 32.0                           3128   
3          64.0                  3.0                         179757   
4        4707.0                 11.0                           1147   

   total_deaths_per_1m_population  total_tests  total_tests_per_1m_population  \
0                           125.0     612112.0                        15381.0   
1                           854.0     805546.0                       280223.0   
2                            83.0     230861.0                         5172.0   
3                          1641.0     193595.0                      2501615.0   
4                            27.0     648883.0                        19154.0   

   population  
0    39797047  
1     2874665  
2    44636630  
3       77388  
4    33876821

In [7]:
dfCS.count()

country                           220
continent                         220
total_confirmed                   220
total_deaths                      207
total_recovered                   219
active_cases                      219
serious_or_critical               148
total_cases_per_1m_population     220
total_deaths_per_1m_population    203
total_tests                       210
total_tests_per_1m_population     210
population                        220
dtype: int64

In [8]:
filepath4 = '../Data/worldometer_coronavirus_daily_data.csv'
dfDaily = pd.read_csv(filepath4)
dfDaily.head()

date      country  cumulative_total_cases  daily_new_cases  \
0  2020-2-15  Afghanistan                     0.0              NaN   
1  2020-2-16  Afghanistan                     0.0              NaN   
2  2020-2-17  Afghanistan                     0.0              NaN   
3  2020-2-18  Afghanistan                     0.0              NaN   
4  2020-2-19  Afghanistan                     0.0              NaN   

   active_cases  cumulative_total_deaths  daily_new_deaths  
0           0.0                      0.0               NaN  
1           0.0                      0.0               NaN  
2           0.0                      0.0               NaN  
3           0.0                      0.0               NaN  
4           0.0                      0.0               NaN

In [9]:
dfDaily.count()

date                       110464
country                    110464
cumulative_total_cases     110464
daily_new_cases            103591
active_cases               109962
cumulative_total_deaths    103938
daily_new_deaths            89375
dtype: int64

#### notes
world happiness 2021 and summary of covid cases tables I can join to hopefully find some interesting things. World happiness over the years I can look at and see if there is a dip this last year in culumative scores. Daily covid cases is not as useful to me. Here are my data sources https://www.kaggle.com/josephassaker/covid19-global-dataset?select=worldometer_coronavirus_summary_data.csv https://www.kaggle.com/ajaypalsinghlo/world-happiness-report-2021 

## Try a Join to see how much Data is left

In [10]:
df21.rename(columns={'Country name':'country'}, inplace=True)

In [11]:
df21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149 entries, 0 to 148
Data columns (total 20 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   country                                     149 non-null    object 
 1   Regional indicator                          149 non-null    object 
 2   Ladder score                                149 non-null    float64
 3   Standard error of ladder score              149 non-null    float64
 4   upperwhisker                                149 non-null    float64
 5   lowerwhisker                                149 non-null    float64
 6   Logged GDP per capita                       149 non-null    float64
 7   Social support                              149 non-null    float64
 8   Healthy life expectancy                     149 non-null    float64
 9   Freedom to make life choices                149 non-null    float64
 10  Generosity    

In [12]:
dfCS.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 220 entries, 0 to 219
Data columns (total 12 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   country                         220 non-null    object 
 1   continent                       220 non-null    object 
 2   total_confirmed                 220 non-null    int64  
 3   total_deaths                    207 non-null    float64
 4   total_recovered                 219 non-null    float64
 5   active_cases                    219 non-null    float64
 6   serious_or_critical             148 non-null    float64
 7   total_cases_per_1m_population   220 non-null    int64  
 8   total_deaths_per_1m_population  203 non-null    float64
 9   total_tests                     210 non-null    float64
 10  total_tests_per_1m_population   210 non-null    float64
 11  population                      220 non-null    int64  
dtypes: float64(7), int64(3), object(2)
m

In [13]:
df = df21.merge(dfCS, on='country')
df.head()

country Regional indicator  Ladder score  \
0      Finland     Western Europe         7.842   
1      Denmark     Western Europe         7.620   
2  Switzerland     Western Europe         7.571   
3      Iceland     Western Europe         7.554   
4  Netherlands     Western Europe         7.464   

   Standard error of ladder score  upperwhisker  lowerwhisker  \
0                           0.032         7.904         7.780   
1                           0.035         7.687         7.552   
2                           0.036         7.643         7.500   
3                           0.059         7.670         7.438   
4                           0.027         7.518         7.410   

   Logged GDP per capita  Social support  Healthy life expectancy  \
0                 10.775           0.954                     72.0   
1                 10.933           0.954                     72.7   
2                 11.117           0.942                     74.4   
3                 10.878           0.983                     73.0   
4                 10.932           0.942                     72.4   

   Freedom to make life choices  ...  total_confirmed  total_deaths  \
0                         0.949  ...            95742         969.0   
1                         0.946  ...           293677        2534.0   
2                         0.919  ...           703004       10891.0   
3                         0.955  ...             6649          30.0   
4                         0.913  ...          1685000       17745.0   

   total_recovered  active_cases  serious_or_critical  \
0          46000.0       48773.0                  7.0   
1         288190.0        2953.0                 12.0   
2         683865.0        8248.0                 41.0   
3           6596.0          23.0                  1.0   
4        1631715.0       35540.0                137.0   

   total_cases_per_1m_population  total_deaths_per_1m_population  total_tests  \
0                          17253                           175.0    5420876.0   
1                          50525                           436.0   68626438.0   
2                          80642                          1249.0    8306968.0   
3                          19360                            87.0     767476.0   
4                          98121                          1033.0   14730069.0   

   total_tests_per_1m_population  population  
0                       976859.0     5549294  
1                     11806675.0     5812512  
2                       952897.0     8717589  
3                      2234627.0      343447  
4                       857764.0    17172634  

[5 rows x 31 columns]

In [14]:
df.count()

country                                       136
Regional indicator                            136
Ladder score                                  136
Standard error of ladder score                136
upperwhisker                                  136
lowerwhisker                                  136
Logged GDP per capita                         136
Social support                                136
Healthy life expectancy                       136
Freedom to make life choices                  136
Generosity                                    136
Perceptions of corruption                     136
Ladder score in Dystopia                      136
Explained by: Log GDP per capita              136
Explained by: Social support                  136
Explained by: Healthy life expectancy         136
Explained by: Freedom to make life choices    136
Explained by: Generosity                      136
Explained by: Perceptions of corruption       136
Dystopia + residual                           136


#### notes
Thats not to bad, only lost 13 countries in the merge. There were alot more than that in the covid set but I am only counting the ones that would have been there in the WHR set. Now I just need to decide which columns to keep before exporting it.

In [15]:
df.dropna().count()

country                                       108
Regional indicator                            108
Ladder score                                  108
Standard error of ladder score                108
upperwhisker                                  108
lowerwhisker                                  108
Logged GDP per capita                         108
Social support                                108
Healthy life expectancy                       108
Freedom to make life choices                  108
Generosity                                    108
Perceptions of corruption                     108
Ladder score in Dystopia                      108
Explained by: Log GDP per capita              108
Explained by: Social support                  108
Explained by: Healthy life expectancy         108
Explained by: Freedom to make life choices    108
Explained by: Generosity                      108
Explained by: Perceptions of corruption       108
Dystopia + residual                           108


#### notes
The 108 countries with all the columns should be good gives me a lot to work with.

In [16]:
df.rename(columns={'Regional indicator':'region', 'Ladder score': 'ladder_score',
                   'Standard error of ladder score': 'std_err', 'Logged GDP per capita': 'GDP_per_capita', 
                  'Social support': 'social_support', 'Healthy life expectancy': 'healthy_life_expectancy',
                  'Freedom to make life choices': 'freedom_to_make_life_choices',
                  'Generosity': 'generosity', 'Perceptions of corruption': 'perceptions_of_corruption',
                  'Ladder score in Dystopia': 'ladder_score_in_dystopia',
                  'Explained by: Log GDP per capita': 'eb_GDP_per_capita',
                  'Explained by: Social support': 'eb_social_support',
                  'Explained by: Healthy life expectancy': 'eb_healthy_life_expectancy',
                  'Explained by: Freedom to make life choices': 'eb_freedom_to_make_life_choices',
                  'Explained by: Generosity': 'eb_generosity',
                  'Explained by: Perceptions of corruption': 'eb_perceptions_of_corruption',
                  'Dystopia + residual': 'dystopia_plus_residual'}, inplace=True)

In [17]:
df.dropna(inplace=True)
df.count()

country                            108
region                             108
ladder_score                       108
std_err                            108
upperwhisker                       108
lowerwhisker                       108
GDP_per_capita                     108
social_support                     108
healthy_life_expectancy            108
freedom_to_make_life_choices       108
generosity                         108
perceptions_of_corruption          108
ladder_score_in_dystopia           108
eb_GDP_per_capita                  108
eb_social_support                  108
eb_healthy_life_expectancy         108
eb_freedom_to_make_life_choices    108
eb_generosity                      108
eb_perceptions_of_corruption       108
dystopia_plus_residual             108
continent                          108
total_confirmed                    108
total_deaths                       108
total_recovered                    108
active_cases                       108
serious_or_critical      

In [18]:
fileout = '../Data/happyVcovid.csv'
df.to_csv(fileout, index=False)